File: ekodb-client-ts/src/join.ts
/**
 * Join support for multi-collection queries
 *
 * This module provides support for joining data across multiple collections,
 * similar to SQL joins but with document-oriented semantics.
 */

/**
 * Configuration for joining collections
 */
export interface JoinConfig {
  /** Target collections to join with */
  collections: string[];
  /** Field in the current collection */
  local_field: string;
  /** Field in the target collection(s) */
  foreign_field: string;
  /** Name of the field to store joined data */
  as_field: string;
}

/**
 * Builder for constructing join configurations
 *
 * @example
 * ```typescript
 * // Single collection join
 * const join = JoinBuilder.single("users", "user_id", "id", "user");
 *
 * // Multi-collection join
 * const join = JoinBuilder.multi(
 *   ["users", "profiles", "settings"],
 *   "user_id",
 *   "id",
 *   "user_info"
 * );
 *
 * // Use in query
 * const query = new QueryBuilder()
 *   .eq("status", "active")
 *   .join(join)
 *   .build();
 * ```
 */
export class JoinBuilder {
  private config: JoinConfig;

  private constructor(
    collections: string[],
    localField: string,
    foreignField: string,
    asField: string,
  ) {
    this.config = {
      collections,
      local_field: localField,
      foreign_field: foreignField,
      as_field: asField,
    };
  }

  /**
   * Create a join with a single collection
   *
   * @param collection - Target collection name
   * @param localField - Field in the current collection
   * @param foreignField - Field in the target collection
   * @param asField - Name of the field to store joined data
   */
  static single(
    collection: string,
    localField: string,
    foreignField: string,
    asField: string,
  ): JoinConfig {
    return new JoinBuilder(
      [collection],
      localField,
      foreignField,
      asField,
    ).build();
  }

  /**
   * Create a join with multiple collections
   *
   * @param collections - Array of target collection names
   * @param localField - Field in the current collection
   * @param foreignField - Field in the target collections
   * @param asField - Name of the field to store joined data
   */
  static multi(
    collections: string[],
    localField: string,
    foreignField: string,
    asField: string,
  ): JoinConfig {
    return new JoinBuilder(
      collections,
      localField,
      foreignField,
      asField,
    ).build();
  }

  /**
   * Build the final JoinConfig
   */
  build(): JoinConfig {
    return this.config;
  }
}

File: ekodb-client-py/python/ekodb_client/join.py
"""
JoinConfig for cross-collection joins in ekoDB queries.

Example:
    ```python
    from ekodb_client import QueryBuilder, JoinConfig

    join = JoinConfig.single("users", "user_id", "id", "user")

    query = (
        QueryBuilder()
        .eq("status", "active")
        .join(join.to_dict())
        .build()
    )

    results = await client.find("orders", query)
    ```
"""

from typing import Any, Dict, List


class JoinConfig:
    """Configuration for joining data across collections."""

    def __init__(
        self,
        collections: List[str],
        local_field: str,
        foreign_field: str,
        as_field: str,
    ):
        """Create a join configuration.

        Args:
            collections: Target collections to join with
            local_field: Field in the source collection
            foreign_field: Field in the target collection(s)
            as_field: Name of the field to store joined data
        """
        self.collections = collections
        self.local_field = local_field
        self.foreign_field = foreign_field
        self.as_field = as_field

    @classmethod
    def single(
        cls, collection: str, local_field: str, foreign_field: str, as_field: str
    ) -> "JoinConfig":
        """Create a join with a single collection."""
        return cls([collection], local_field, foreign_field, as_field)

    @classmethod
    def multi(
        cls,
        collections: List[str],
        local_field: str,
        foreign_field: str,
        as_field: str,
    ) -> "JoinConfig":
        """Create a join with multiple collections."""
        return cls(collections, local_field, foreign_field, as_field)

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dict for use in queries."""
        return {
            "collections": self.collections,
            "local_field": self.local_field,
            "foreign_field": self.foreign_field,
            "as_field": self.as_field,
        }
